### ISSUE TITLE
Cannot deploy packages anymore

### ISSUE THREAD

**Issue by RobinBreman** [SharePoint/sp-dev-docs#2624]
#### Category
- [ ] Question
- [ ] Typo
- [x ] Bug
- [ ] Additional article idea

Cannot deploy a previously good package anymore, this now results in a strange app catalog behavior.

#### Observed Behavior

Grabbed a previously good package from an earlier build, deployed this package to the app catalog of a fresh (first release enabled) tenant.

![2018-09-24_1614](https://user-images.githubusercontent.com/25765614/45957837-45f07480-c016-11e8-9bfa-3abd6155f7c4.png)

- The Title and ProductId are missing, and the version is incorrectly shown.
- There is no deployment error logged.

![2018-09-24_1616](https://user-images.githubusercontent.com/25765614/45957844-4a1c9200-c016-11e8-9dc1-618a7bfc9d67.png)

- App details screen is showing: 'Invalid SharePoint App Product ID:'

**Comment by RobinBreman** [comment]
**Update**

By dismantling our project and keep digging down I found the cause of this one.

My `webApiPermissionRequests` object has grown in time, this is causing the issue.

When the (formatted) _string_-length of the  webApiPermissionRequests object exceeds 255 chars the solution will not deploy. There is no error shown in the appcatalog only an invalid app is listed. The 'trust solution' dialog is never shown.

The appcatalog library's WebApiPermissionScopes field is the root cause. I found that on uploading the `webApiPermissionRequests` are stored into the hidden WebApiPermissionScopes field. This field cannot exceed 255 chars!
Values are stored in the: `resource, scope; resource,scope ...` format.

You can check the length:
```js
let package = {}; //paste package-solution.json here

package.solution.webApiPermissionRequests.map(function (elem) {
    return elem.resource + ', ' + elem.scope;
}).join("; ").length
```
If this check exceeds 255 chars you will not be able to correctly upload a *new* app-package.

Make sure to **delete** an earlier version of the app-package from the app catalog to experience this behavior.
(If you do _update_ the app you will notice the version and other properties do not change but the 'trust solution' dialog does popup).

**Comment by RobinBreman** [comment]
Dear @VesaJuvonen  has the app catalog deployment changed since v 1.6 ?

**Comment by PzYon** [comment]
Please also see #2642, it's going in the same direction. However it stopped working, without any changes in the app/manifest. From my understanding SP must have changed something on their side, but not together with the release of 1.6, but in the last couple of days. Either some permission `ResourceId`s are suddenly not supported anymore or the handling of invalid permission requests has been changed.

**Comment by martinvillysson** [comment]
I'm also seeing the same problem when trying to upload the starter kit [https://github.com/SharePoint/sp-starter-kit](https://github.com/SharePoint/sp-starter-kit)

**Comment by justdevelopment** [comment]
Looking at the code for the starter-kit it's also because the combined scopes string is to long. 

Our current workaround is using multiple solutions which set scopes (so each one will stay below the 255 limit, as checked by @RobinBreman 's script). You can then deploy and approve all of them, the main solution which actually uses the scopes will work correctly after.

**Comment by PzYon** [comment]
@justdevelopment yes, you are right. I just tested with the permission scopes I mentioned in the other issue and it works - if there aren't too many of them... ;) In other words, I'll close #2642.

Is there any news from Microsoft on why they introduced a character limitation? And how to solve this? Creating multiple Apps just to request new permissions doesn't feel right.

**Comment by PzYon** [comment]
@VesaJuvonen, @patmill Do have an answer for this? I.e.:
- Will this be fixed? If yes, is there a timeline?
- Or do we really need to create multiple apps, just to register permissions?
- Or is there another way to workaround this issue?

My standpoint is clear: We have an app retrieving data from Graph, SP as well as custom APIs, so we hit the 255 character limitation pretty fast. From a developer as well as from a customer perspective, it doesn't make sense to create and maintain "dummy" applications just to be able to register permissions.

**Comment by MatthiasWeibel** [comment]
@VesaJuvonen, @patmill Any updates on this?
